### PR TITLE
SB-1146: Rework `RepositoryFileAttributes.isArtifact()` attribute logic. [in-progress]

### PR DESCRIPTION
**Pull request by sbespalov**
Also there will be `LayoutProvider` and artifact resolution improvements within this PR

### CODE REVIEW

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java [hunk 1]
```diff
@@ -55,7 +55,7 @@ protected boolean isMetadata(Path path)
     {
         try
         {
-            return Boolean.TRUE.equals(RepositoryFiles.isMetadata((RepositoryPath) path));
+            return Boolean.TRUE.equals(RepositoryFiles.isLsm((RepositoryPath) path));
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java]: @sbespalov : What is `isLsm(...)`? What doest this abbreviation stand for?

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java]: it's Layout Specific Metadata, which probably can be better named.

Comment (carlspring) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java]: I think you should just call it `isLayoutSpecificMetadata()` in this case.

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java]: ok, I just thought that such name will be too long :)

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/AbstractArtifactLocationHandler.java]: @carlspring, while working with this PR I realized that we don't need such attribute. There will be enough just `isArtifact` and `isMetada` attributes, which can be overriden in concrete `LayoutProvider` implementation. For example in case of maven we will have following:
- `isMetadata = .pom || maven-metadata.xml`
- `isArtifact = !maven-metadata.xml && !index && !checksum && !diractory`

So we will store `ArtifactEntry` in case of just `isArtifact`, and for example we can detect `maven-metadata.xml` with `!isArtifact && isMetadata`.

File: strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java [hunk 2]
```diff
@@ -33,13 +31,15 @@ public AbstractArtifactCoordinates(Map<String, String> coordinates)
         this.path = toPath();
     }
     
-    protected final void defineCoordinates(String... coordinates)
+    protected final void resetCoordinates(String... coordinates)
```
Comment (carlspring) [inline strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java]: Why is this renamed?

File: strongbox-cron/strongbox-cron-tasks/src/main/java/org/carlspring/strongbox/cron/jobs/RegenerateChecksumCronJob.java [hunk 3]
```diff
@@ -75,9 +75,9 @@ private void regenerateRepositoriesChecksum(String storageId,
     {
         Map<String, Repository> repositories = getRepositories(storageId);
 
-        for (String repository : repositories.keySet())
+        for (String repositoryId : repositories.keySet())
```
Comment (carlspring) [inline strongbox-cron/strongbox-cron-tasks/src/main/java/org/carlspring/strongbox/cron/jobs/RegenerateChecksumCronJob.java]: :+1:

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/config/HazelcastConfiguration.java [hunk 4]
```diff
@@ -18,13 +21,24 @@
 public class HazelcastConfiguration
 {
 
+    public static final int ARTIFACT_ENTRY_CACHE_INVALIDATE_INTERVAL = 60;
+    
+    public static MapConfig nearCacheConfig(String name)
+    {
+        return new MapConfig().setName(name).setNearCacheConfig(new NearCacheConfig().setCacheLocalEntries(true)
```
Comment (carlspring) [inline strongbox-data-service/src/main/java/org/carlspring/strongbox/config/HazelcastConfiguration.java]: Just for my better understanding, what is this near cache config used for?

File: strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java (hunk 2 above)
Comment (sbespalov) [reply strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java]: my thoughts was that this method reset all coordinates to `null`, so it's not **set** or **define** but more like  **reset**. 
I can return it back if you don't like this name.

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/config/HazelcastConfiguration.java (hunk 4 above)
Comment (sbespalov) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/config/HazelcastConfiguration.java]: `NearCache` is fast cache which is not shared between nodes. It used to speed up `ArtifactEntry` fetches within HTTP request, if we fetch it more then once within current request.

File: strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java (hunk 2 above)
Comment (carlspring) [reply strongbox-commons/src/main/java/org/carlspring/strongbox/artifact/coordinates/AbstractArtifactCoordinates.java]: (I was just trying to understand what's goign on :) ).

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/config/HazelcastConfiguration.java (hunk 4 above)
Comment (carlspring) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/config/HazelcastConfiguration.java]: Thanks for the explanation! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/artifact/MavenRepositoryArtifact.java [hunk 5]
```diff
@@ -42,22 +51,22 @@ public MavenDetachedArtifact(final String groupId,
...
-    public MavenDetachedArtifact(final String groupId,
+    public MavenRepositoryArtifact(final String groupId,
                                  final String artifactId,
                                  final String version)
     {
         super(groupId, artifactId, version);
     }
 
-    public MavenDetachedArtifact(final String groupId,
+    public MavenRepositoryArtifact(final String groupId,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/artifact/MavenRepositoryArtifact.java]: So, what is a `MavenRepositoryArtifact` and how is it different to the `MavenDetachedArtifact`...?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/MavenFileSystem.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+import org.carlspring.strongbox.providers.io.RepositoryFileSystem;
+import org.carlspring.strongbox.providers.io.RepositoryFileSystemProvider;
+import org.carlspring.strongbox.storage.repository.Repository;
+
+/**
+ * @author sbespalov
+ *
+ */
+public class MavenFileSystem extends RepositoryFileSystem
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/MavenFileSystem.java]: Cool! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/MavenRepositoryManagementServiceImplTest.java [hunk 7]
```diff
@@ -56,10 +58,10 @@ public static void cleanUp()
     public static Set<MutableRepository> getRepositoriesToClean()
     {
         Set<MutableRepository> repositories = new LinkedHashSet<>();
-        repositories.add(createRepositoryMock(STORAGE0, REPOSITORY_RELEASES_1));
-        repositories.add(createRepositoryMock(STORAGE0, REPOSITORY_RELEASES_2));
-        repositories.add(createRepositoryMock(STORAGE0, REPOSITORY_RELEASES_MERGE_1));
-        repositories.add(createRepositoryMock(STORAGE0, REPOSITORY_RELEASES_MERGE_2));
+        repositories.add(createRepositoryMock(STORAGE0, REPOSITORY_RELEASES_1, Maven2LayoutProvider.ALIAS));
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/MavenRepositoryManagementServiceImplTest.java]: For every layout provider there's a `RepositoryFactory` implementation. (Here it would be `MavenRepositoryFactory`).
I think it would be better to use it instead...?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/storage/indexing/StrongboxIndexerTest.java [hunk 8]
```diff
@@ -81,7 +83,7 @@ public static void cleanUp()
     public static Set<MutableRepository> getRepositoriesToClean()
     {
         Set<MutableRepository> repositories = new LinkedHashSet<>();
-        repositories.add(createRepositoryMock(STORAGE0, REPOSITORY_RELEASES_1));
+        repositories.add(createRepositoryMock(STORAGE0, REPOSITORY_RELEASES_1, Maven2LayoutProvider.ALIAS));
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/storage/indexing/StrongboxIndexerTest.java]: There's a whole bunch of these...

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/MavenRepositoryManagementServiceImplTest.java (hunk 7 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/MavenRepositoryManagementServiceImplTest.java]: maybe, but I was just fixed legacy code here and nothing more

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/MavenRepositoryManagementServiceImplTest.java]: if you think it need to be improved, maybe OSS guys can help with this?

**Comment by carlspring** [comment]
Great work! Thanks! :)
